pkg/debug/debug.go

```
@@ -0,0 +1,144 @@
+	if artifactsDirectoryBase == "" {
+		return fmt.Errorf("$%s not set", testArtifactsDirectoryVarName)
+	}
+
+	_, _ = fmt.Fprintf(writer, "collecting namespaced resources for debugging...\n")
+
+	cmd := execCommand(
+		kubectlPath,
+		"--kubeconfig",
+		kubeConfigPath,
+		"api-resources",
+		"--verbs=list",
+		"--namespaced=true",
+		"-o",
+		"name")
+	cmd.Stderr = writer
+
+	output, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("fetching API resource types failed: %v", err)
+	}
+
+	artifactsDirectory := path.Join(artifactsDirectoryBase, namespace)
+
+	err = fs.MkdirAll(artifactsDirectory, 0777)
+	if err != nil {
+		return fmt.Errorf("creating %q failed: %v", artifactsDirectory, err)
+	}
+
+	resources := strings.Split(string(output), "\n")
+	groupedResources := make(map[string][]string)
+
+	for _, resources := range resources {
+		if resources == "" {
+			continue
+		}
+
+		resourcesAndGroup := strings.SplitN(resources, ".", 2)
+
+		var group string
+
+		if len(resourcesAndGroup) == 2 {
+			group = resourcesAndGroup[1]
+		}
+
+		groupedResources[group] = append(groupedResources[group], resources)
+	}
+
+	var wg sync.WaitGroup
+
+	for group, resourcesSlice := range groupedResources {
+		sort.Strings(resourcesSlice)
+
+		resources := strings.Join(resourcesSlice, ",")
+
+		var fileName string
+
+		if group == "" {
+			fileName = "resources.yaml"
+		} else {
+			fileName = fmt.Sprintf("resources-%s.yaml", group)
+		}
+
+		wg.Add(1)
+
+		go collectResources(fs, writer, &wg, namespace, fileName, resources,
+			artifactsDirectory, kubectlPath, kubeConfigPath)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func collectResources(fs afero.Fs, writer io.Writer, wg *sync.WaitGroup, namespace, fileName, resourcesNames,
+	directoryName, kubectlPath, kubeConfigPath string) {
+	defer wg.Done()
+
+	outPath := path.Join(directoryName, fileName)
+
+	output, err := fs.Create(outPath)
+	if err != nil {
+		_, _ = fmt.Fprintf(writer, "creating %q failed: %v\n", outPath, err)
```

Why not return an error here? Same for the `fmt.Fprintf` below.

---

See point 1 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384